src/libopensc/card-esteid2025.c

```diff
@@ -0,0 +1,224 @@
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		r = iso_ops->process_fci(card, file, resp, resplen);
+		LOG_TEST_RET(card->ctx, r, "Process fci failed");
+		*file_out = file;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+esteid_set_security_env(sc_card_t *card, const sc_security_env_t *env, int se_num)
+{
+	struct sc_apdu apdu;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (env == NULL || env->key_ref_len != 1)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	sc_log(card->ctx, "algo: %lu operation: %d keyref: %d", env->algorithm, env->operation, env->key_ref[0]);
+
+	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN) {
+		const u8 cse_crt_sig[] = {0x80, 0x01, 0x54, 0x84, 0x01, env->key_ref[0]};
+		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xB6, cse_crt_sig, sizeof(cse_crt_sig), NULL, 0);
+	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_DERIVE) {
+		const u8 cse_crt_der[] = {0x84, 0x01, env->key_ref[0]};
+		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xB8, cse_crt_der, sizeof(cse_crt_der), NULL, 0);
+	} else {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	}
+	SC_TRANSMIT_TEST_RET(card, apdu, "SET SECURITY ENV failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+esteid_compute_signature(sc_card_t *card, const u8 *data, size_t datalen, u8 *out, size_t outlen)
+{
+	struct sc_apdu apdu;
+	u8 sbuf[SIGNATURE_PAYLOAD_SIZE + 2] = {0x90, SIGNATURE_PAYLOAD_SIZE};
+	size_t le = MIN(SC_MAX_APDU_RESP_SIZE, MIN(SIGNATURE_PAYLOAD_SIZE * 2, outlen));
+
+	LOG_FUNC_CALLED(card->ctx);
+	if (data == NULL || out == NULL || datalen > SIGNATURE_PAYLOAD_SIZE)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	// left-pad if necessary
+	memcpy(&sbuf[SIGNATURE_PAYLOAD_SIZE + 2 - datalen], data, MIN(datalen, SIGNATURE_PAYLOAD_SIZE));
+	datalen = SIGNATURE_PAYLOAD_SIZE + 2;
+
+	sc_format_apdu_ex(&apdu, 0x00, 0x2A, 0x90, 0xA0, sbuf, datalen, NULL, 0);
+	SC_TRANSMIT_TEST_RET(card, apdu, "PSO Set Hash failed");
+
+	sc_format_apdu_ex(&apdu, 0x00, 0x2A, 0x9E, 0x9A, NULL, 0, out, le);
+	SC_TRANSMIT_TEST_RET(card, apdu, "PSO Compute Digital Signature failed");
+
+	LOG_FUNC_RETURN(card->ctx, (int)apdu.resplen);
+}
+
+static int
+esteid_get_pin_remaining_tries(sc_card_t *card, int pin_reference)
+{
+	const u8 get_pin_info[] = {0xA0, 0x03, 0x83, 0x01, pin_reference};
+	struct sc_apdu apdu;
+	u8 apdu_resp[SC_MAX_APDU_RESP_SIZE];
+	size_t taglen;
+	LOG_FUNC_CALLED(card->ctx);
+
+	sc_format_apdu_ex(&apdu, 0x00, 0xCB, 0x00, 0xFF, get_pin_info, sizeof(get_pin_info), apdu_resp, sizeof(apdu_resp));
+	SC_TRANSMIT_TEST_RET(card, apdu, "GET DATA(pin info) failed");
+	if (apdu.resplen < 3 || apdu.resp[0] != 0xA0)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	const u8 *tag = sc_asn1_find_tag(card->ctx, apdu_resp + 2, apdu.resplen - 2, 0xDF21, &taglen);
+	if (tag == NULL)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	return tag[0];
```

Comment:
I think you need to check for `taglen >= 1`

Reply:
Fixed